controllers/vspherecluster_controller.go

```diff
@@ -170,41 +166,62 @@ func (r *VSphereClusterReconciler) reconcileAPIEndpoints(ctx *context.ClusterCon
 		return nil
 	}
 
-	// Get the oldest control plane machine in the cluster.
-	machine, err := infrautilv1.GetOldestControlPlaneMachine(
-		ctx, ctx.Client, ctx.VSphereCluster.Namespace, ctx.VSphereCluster.Name)
+	// Get the CAPI Machine resources for the cluster.
+	machines, err := infrautilv1.GetMachinesInCluster(ctx, ctx.Client, ctx.VSphereCluster.Namespace, ctx.VSphereCluster.Name)
 	if err != nil {
-		return errors.Wrapf(err,
-			"failed to get oldest control plane machine for VSphereCluster %s/%s",
+		errors.Wrapf(err,
+			"failed to get Machinces for Cluster %s/%s",
 			ctx.VSphereCluster.Namespace, ctx.VSphereCluster.Name)
 	}
 
-	// Get the oldest control plane machine's VSphereMachine resource.
-	vsphereMachine, err := infrautilv1.GetVSphereMachine(
-		ctx, ctx.Client, ctx.VSphereCluster.Namespace, machine.Name)
-	if err != nil {
-		return errors.Wrapf(err,
-			"failed to get VSphereMachine for Machine %s/%s/%s",
-			ctx.VSphereCluster.Namespace, ctx.VSphereCluster.Name, machine.Name)
-	}
+	// Iterate over the cluster's control plane CAPI machines.
+	for _, machine := range clusterutilv1.GetControlPlaneMachines(machines) {
 
-	// Get the machine's preferred IP address.
-	ipAddr, err := infrautilv1.GetMachinePreferredIPAddress(vsphereMachine)
-	if err != nil {
-		return errors.Wrapf(err,
-			"failed to get preferred IP address for VSphereMachine %s/%s/%s",
-			ctx.VSphereCluster.Namespace, ctx.VSphereCluster.Name, vsphereMachine.Name)
-	}
+		// Only machines with bootstrap data will have an IP address.
+		if machine.Spec.Bootstrap.Data == nil {
+			ctx.Logger.V(6).Info(
+				"skipping machine while looking for IP address",
+				"machine-name", machine.Name,
+				"skip-reason", "nilBootstrapData")
+			continue
+		}
 
-	// Set APIEndpoints so the Cluster API Cluster Controller can pull them
-	ctx.VSphereCluster.Status.APIEndpoints = []infrav1.APIEndpoint{
-		{
-			Host: ipAddr,
-			Port: apiEndpointPort,
-		},
+		// Get the VSphereMachine for the CAPI Machine resource.
+		vsphereMachine, err := infrautilv1.GetVSphereMachine(ctx, ctx.Client, machine.Namespace, machine.Name)
```

Comment:
Hi @yastij,

Can you first please explain how to avoid this? Because the logic here involves first getting members of the control plane (annotations only set on the CAPI `Machine` resources) and then getting the IP address(es) from the `VSphereMachine` resources.

Either way, I don't want to modify this in this PR. I'd rather get this merged and allow you to iterate.

Reply:
The idea is to make `GetVSphereMachines()` return a map and use the machine name as a key. then we can call `GetControlPlaneMachines()` and  `GetVsphereMachines()`,  iterate over `GetControlPlaneMachines()` and use the name to retrieve the vsphereMachine. 

This would avoid making multiple requests to the apiserver, and anyway Listing is done against a cache. I'm fine with leaving as Is and iterate If you want to merge this quickly

Reply:
I like the idea, but yeah, let's leave this for a follow-up.

Reply:
> If you want to merge this quickly

I want `master` to work. This makes `master` work :)